Lib/asyncio/base_events.py

```diff
@@ -1959,28 +1962,32 @@ def _run_once(self):
             # Remove delayed calls that were cancelled if their number
             # is too high
             new_scheduled = []
-            for handle in self._scheduled:
+            for when_handle in self._scheduled:
+                handle = when_handle[1]
                 if handle._cancelled:
                     handle._scheduled = False
                 else:
-                    new_scheduled.append(handle)
+                    new_scheduled.append(when_handle)
 
             heapq.heapify(new_scheduled)
             self._scheduled = new_scheduled
             self._timer_cancelled_count = 0
         else:
             # Remove delayed calls that were cancelled from head of queue.
-            while self._scheduled and self._scheduled[0]._cancelled:
+            while self._scheduled and self._scheduled[0][1]._cancelled:
                 self._timer_cancelled_count -= 1
-                handle = heapq.heappop(self._scheduled)
+                _, handle = heapq.heappop(self._scheduled)
                 handle._scheduled = False
 
         timeout = None
-        if self._ready or self._stopping:
+        ready = self._ready
+        scheduled = self._scheduled
```

Comment:
I think you can just forget about those optimizations (see my comment). Maybe if you give me benchmarks I'd be more convinced but otherwise I don't think accessing the attributes is actually costly (even though there might be a lot of calls).

Reply:
Some prior similar art

#106751
#106864
#106754